File: lang_go/token.go
package main

type TokenType int

const (
	Invalid TokenType = iota
	Identifier
	Number
	Colon
	Comma
	Period
	LBracket
	RBracket
	LBrace
	RBrace
	Plus
	Minus
	Equals
	Times
	Int
	Float
	NewLine
	Tab
	EndOfTokens
)

func (t TokenType) String() string {
	return map[TokenType]string{
		Identifier: "Identifier",
		Number:     "Number",
		Colon:      "Colon",
		Comma:      "Comma",
		Period:     "Period",
		LBracket:   "LBracket",
		RBracket:   "RBracket",
		LBrace:     "LBrace",
		RBrace:     "RBrace",
		Plus:       "Plus",
		Minus:      "Minus",
		Equals:     "Equals",
		Times:      "Times",
		Int:        "Int",
		Float:      "Float",
		NewLine:    "NewLine",
		Tab:        "Tab",
		Invalid:    "Invalid",
		EndOfTokens: "EndOfTokens",
	}[t]
}

var mapChar = map[string]TokenType{
	":":  Colon,
	",":  Comma,
	".":  Period,
	"[":  LBracket,
	"]":  RBracket,
	"{":  LBrace,
	"}":  RBrace,
	"+":  Plus,
	"-":  Minus,
	"=":  Equals,
	"*":  Times,
	"\n": NewLine,
	"\t": Tab,
}

var mapWord = map[string]TokenType{
	"int":   Int,
	"float": Float,
}

type Token struct {
	Type  TokenType
	Value string
}

func init() {
	go getTokens()
}

var tokens = make(chan Token)

func getTokens() {
	for {
		lex, ok := <-lexemes
		if !ok {
			break
		}

		switch lex.Type {
		case lexemeChar:
			if tokType, ok := mapChar[lex.Value]; ok {
				tokens <- Token{Type: tokType}
			} else {
				tokens <- Token{Type: Invalid, Value: lex.Value}
			}

		case lexemeWord:
			if tokType, ok := mapWord[lex.Value]; ok {
				tokens <- Token{Type: tokType}
			} else {
				tokens <- Token{Type: Identifier, Value: lex.Value}
			}

		case lexemeNumber:
			tokens <- Token{
				Type:  Number,
				Value: lex.Value,
			}
		}
	}
	tokens <- Token{ Type:  EndOfTokens }

	close(tokens)
}

File: lang/token.c
#include <stdio.h>
#include "token.h"
#include "lexeme.h"
#include "strings.h"

Token getNextToken() {
	Lexeme lex = getNextLexeme();
	Token tok = {type: EndOfTokens, value: NULL};

	switch (lex.type) {
	case lexemeChar:
		switch (lex.value[0]) {
		case '+':
			tok.type = Plus;
			break;
		case ':':
			tok.type = Colon;
			break;
		case ';':
			tok.type = Semicolon;
			break;
		case '=':
			tok.type = Equals;
			break;
		case '.':
			tok.type = Period;
			break;
		case ',':
			tok.type = Comma;
			break;
		case '[':
			tok.type = LBracket;
			break;
		case ']':
			tok.type = RBracket;
			break;
		case '\n':
			tok.type = Newline;
			break;
		case '\t':
			tok.type = Tab;
			break;
		case EOF:
			tok.type = EndOfTokens;
			break;
		default:
			fprintf(stderr, "token.c: lexeme char invalid\n");
			tok.type = EndOfTokens;
			break;
		}
		break;

	case lexemeWord:
		if (strcmp(lex.value, "int") == 0) {
			tok.type = Int;
		} else if (strcmp(lex.value, "float") == 0) {
			tok.type = Float;
		} else {
			tok.type = Identifier;
			tok.value = strAlloc(lex.value);
		}
		break;

	case lexemeNumber:
		tok.type = Number;
		tok.value = strAlloc(lex.value);
		break;

	default:
		fprintf(stderr, "token.c: tok.type invalid\n");
		break;
	}

	strFree(&lex.value);
	return tok;
}

void printToken(Token tok) {
	switch (tok.type) {
	case Plus:
		printf("Plus\n");
		break;
	case Semicolon:
		printf("Semicolon\n");
		break;
	case Colon:
		printf("Colon\n");
		break;
	case Equals:
		printf("Equals\n");
		break;
	case Period:
		printf("Period\n");
		break;
	case Comma:
		printf("Comma\n");
		break;
	case RBracket:
		printf("RBracket\n");
		break;
	case LBracket:
		printf("LBracket\n");
		break;
	case Identifier:
		printf("Identifier: %s\n", tok.value);
		break;
	case Number:
		printf("Number    : %s\n", tok.value);
		break;
	case Int:
		printf("Int\n");
		break;
	case Float:
		printf("Float\n");
		break;
	case Newline:
		printf("Newline\n");
		break;
	case Tab:
		printf("Tab\n");
		break;
	case EndOfTokens:
		printf("EndOfTokens\n");
		break;
	default:
		printf("token<Unknown> : %s\n", tok.value);
		break;
	}
}


